src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,217 @@
+        let c = key.encryption_key().encrypt_with(message, scalar);
+        Self(c)
+    }
+}
+
+impl ToString for BlindedMatchKey {
+    fn to_string(&self) -> String {
+        let (c0, c1) = self.0.inner();
+        let mut c0_hex = hex(c0.compress().to_bytes());
+        let c1_hex = hex(c1.compress().to_bytes());
+        c0_hex.push_str(&c1_hex);
+        c0_hex
+    }
+}
+
+struct PrivacyBudget {
+    redis_connection: redis::Connection,
+    total_privacy_budget: u32,
+}
+
+trait ToString {
+    fn to_string(&self) -> String;
+}
+
+impl ToString for Ciphertext {
+    fn to_string(&self) -> String {
+        let (c0, c1) = self.inner();
+        let mut c0_hex = hex(c0.compress().to_bytes());
+        let c1_hex = hex(c1.compress().to_bytes());
+        c0_hex.push_str(&c1_hex);
+        c0_hex
+    }
+}
+
+#[allow(dead_code)]
+impl PrivacyBudget {
+    fn new(
+        redis_host_name: &str,
+        redis_password: Option<String>,
+        is_tls: bool,
+    ) -> Res<PrivacyBudget> {
+        //if Redis server needs secure connection
+        let uri_scheme = if is_tls { "rediss" } else { "redis" };
+
+        let redis_conn_url = format!(
+            "{}://:{}@{}",
+            uri_scheme,
+            redis_password.unwrap_or_default(),
+            redis_host_name
+        );
+
+        let redis_connection = redis::Client::open(redis_conn_url)?.get_connection()?;
+
+        Ok(PrivacyBudget {
+            redis_connection,
+            total_privacy_budget: 100,
+        })
+    }
+
+    /// Input : a vector of ciphertexts alongwith budget to be consumed.
+    /// This atomically increases the budget of all of the blinded match keys that are received.
+    /// For those that end up exceeding the budget, the budget is then decremented again,
+    /// but the corresponding match key is discarded and not passed to the output.
+    pub fn get_eligible_packets_with_budget(
+        &mut self,
+        packets: &[BlindedMatchKey],
+        budget: u32,
+    ) -> Vec<BlindedMatchKey> {
+        let mut eligible_packets: Vec<BlindedMatchKey> = Vec::new();
+        for packet in packets.iter() {
+            let _eligible = match self.increase_budget_consumed(packet, budget) {
+                Ok(true) => {
+                    eligible_packets.push(*packet);
+                    Ok(())
+                }
+                Ok(false) => self.decrease_budget_consumed(&*packet, budget),
+                Err(_) => Ok(()),
+            };
+        }
+        eligible_packets
+    }
+
+    pub fn get_total_privacy_budget(&self) -> u32 {
+        self.total_privacy_budget
+    }
+
+    /// Budget is increased for each key atomically in Redis
+    fn increase_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<bool> {
+        let key = packet.to_string();
+        let (budget_consumed,): (u32,) =
+            transaction(&mut self.redis_connection, &[key.clone()], |con, pipe| {
```

Comment:
Read, write, read seems inefficient.  Can't you just bail if the value you read exceeds the budget?  Also, given that the first read is outside of the atomic transaction, aren't you exposed to concurrent updates?

(I don't know redis well, so feel free to tell me I'm wrong.)

Reply:
It would expose it but I thought we were ok with doing sequential queries for now (based on our discussions in person)?

Reply:
yes, but if we don't care about concurrency, then why use a transaction?